### ISSUE TITLE
Infinite loop hangs compiler / listener / runtime

### ISSUE THREAD

**Issue by catb0t** [factor/factor#1573]
`[ t ] [ ] while` hangs the Listener if typed in interactive mode, making me forcibly kill Factor. It also hangs the compiler (and thus the Listener) if a file containing `[ t ] [ ] while` is `USE:`d or `reload`ed.

I'm aware nobody in their right mind would actually *use* an infinite loop in Factor like one would in a procedural language like C, but I think it's definitely a bug that such a simple expression brings down the runtime.

**Comment by mrjbq7** [comment]
From the command line you can interrupt it using Ctrl-C.

```
$ ./factor
Factor 0.98 x86.64 (1736, heads/master-20883fb21a, Tue Mar  8 07:54:56 2016)
[Clang (GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81))] on macosx
IN: scratchpad [ t ] [ ] while
^CInterrupted
Starting low level debugger...
Basic commands:
  q ^D             -- quit Factor
  c                -- continue executing Factor - NOT SAFE
  t                -- throw exception in Factor - NOT SAFE
  .s .r .c         -- print data, retain, call stacks
  help             -- full help, including advanced commands

> t
Interrupt

Type :help for debugging help.
IN: scratchpad 
```

But from the UI it looks to be not possible?

**Comment by catb0t** [comment]
Yes, from the command-line shell you can use CTRL-C, but I don't use the shell because the Listener is *far* nicer to use.

Perhaps if ALT-C or CTRL-ALT-C did the equivalent of SIGINT to the word running the code typed into the Listener?

**Comment by jonenst** [comment]
On a mac, if you launch factor from the terminal with -run=ui.tools, then can you have the UI listener  while still being able to CTRL-C on the terminal  ?

**Comment by bjourne** [comment]
Yes. Works for me anyhow.

**Comment by catb0t** [comment]
I still think it's silly that any operation actually freezes that whole UI window -- `http.client:http-get` on a slow connection, or iterating on a long file, or anything that takes time, but I guess this is because Factor isn't the best at threading and so the code-runner part of the Listener UI isn't separate.

Is this closed, or...?

**Comment by mrjbq7** [comment]
I don't want to close this, since I think the UI should handle this better...

**Comment by AlexIljin** [comment]
What if we could press Ctrl+Break (on Windows) and have the main thread throw an exception wherever it's currently running? Would that be a good solution? The Listener would then show the error as if the hanging code threw it. Sounds good?

**Comment by bjourne** [comment]
Yes, if you can get the factor process to "hear" that keypress. In a console, ctrl+c works because it is the console that notices the keypress and delivers the sigint signal to factor and we have a handler installed for that. No facility like that exists for gui programs.

**Comment by catb0t** [comment]
Sure, that doesn't sound half-bad. :)

**Comment by AlexIljin** [comment]
![ctrl-break](https://cloud.githubusercontent.com/assets/169039/15629329/0c635604-251f-11e6-8168-b65968a570f4.png)

**Comment by erg** [comment]
Nice photoshop, looks realistic!

**Comment by AlexIljin** [comment]
Yes, it's coming alright. ; )
Interrupting the GC crashes the whole thing, so I'm trying to work around it as we speak.

**Comment by AlexIljin** [comment]
Here's the pull request: #1638. There are still issues, so it's a beta version, as in "beta than nothing".
It's more of a proof of concept, and it desperately needs to be polished by people who know more about the VM and its failure modes. Here's a quote from the main commit:

    The code interrupts busy loops by making the main thread throw the
    STATUS_SINGLE_STEP exception. This does not start the low-level debugger
    mode in the console, but rather returns control to the Listener, which is
    very nice and useful. When the Listener is not in a tight loop, the code
    does nothing (probably the exception is suppressed somewhere by the
    Listener).
    
    Unfortunately, sometimes the whole process crashes. This seems to happen
    when the GC is trapped. It would be good to be able to only interrupt the
    main thread when it is running the user code by analyzing the Eip value so
    that VM code would be exempt from trapping. If that is not feasible, at
    least we can make the GC provide some sort of signal to the Ctrl-Break
    handler that it's not a good time for an interruption.
    
    Another problem with the code is that it doesn't always work, you need to
    activate another window (but not the Factor's console window!) to make it
    react to Ctrl-Break. This is an issue with GetAsyncKeyState and the way
    Windows handles privilege separation. For instance, if Factor is running
    as a non-admin user, and you switch focus to another app with admin
    privileges and press Ctrl-Break, then Factor's thread won't detect that.
    If you do the same thing, but with the other process having the same
    rights as Factor, it will handle the Ctrl-Break event. The funny thing is,
    if the Listener hangs, its main window is replaced with a more responsive
    dummy, which is owned by the Explorer.exe, and that substitution prevents
    Ctrl-Break from working on systems newer than XP, unless you switch to
    another window as described.
    
    Yet another issue is that the keystroke is sensed globally, so if you have
    multiple Factor instances running, they will all get the signal. The good
    news is that only the ones in the tight loop seem to react.

In the text above I was wrong about the Listener ignoring the signal. In fact, if the Listener is *not* in a busy loop, you are very likely to terminate the process instantly. Or you can trap the drawing procedure and have a dead and useless Listener window.

I could not prevent my thread from trapping the GC, even though I tried. It still happens.

It's not a complete solution, but at least you get the idea.

**Comment by bjourne** [comment]
To cleanly stop the main vm thread, you do the fep queue thing:
```
factor_vm* vm = thread_vms.begin()->second;
vm->safepoint.enqueue_fep(vm);
```
That should bring up the low-level debugger in the console and you can then press `t` to abort the endless loop.

**Comment by AlexIljin** [comment]
Can't I just do the "t" thing right away?

**Comment by bjourne** [comment]
No, because your loop is not running on the main thread. You could set a flag on the vm like `vm->ctrl_break_catched_p = true` and in `handle_safepoint` add eg.:

    if (atomic::load(&parent->ctrl_break_catced_p)) {
        parent->ctrl_break_catced_p = false;
        parent->general_error(ERROR_INTERRUPT, parent->false_object, parent->false_object);
    }

But that's details. I don't like your loop in `ctrl_break_thread_proc` because it wakes the thread up once every 10 ms. Really, really bad for performance afaik. And it's not great to catch the ctrl+break if it is not the Factor window that has focus.

**Comment by AlexIljin** [comment]
To fight the problem of the window ghosting that prevents ctrl-break thread from reading the keyboard state I could use the [DisableProcessWindowsGhosting](https://msdn.microsoft.com/en-us/library/ms648415%28v=vs.85%29.aspx) function.

An for performance, seriously, I could do Sleep(1) and it still would not be noticeable. It's just a single function call. On the other hand, Sleep(100) would work just as well, since we humans are not too quick when we push the keyboard. I, for one, can't do more than 10 key presses per second.

And, of course, it is possible to find out which is the foreground window in the system, and whether that window belongs to one of our threads.

I'll try to implement some of these things and see what happens.

**Comment by AlexIljin** [comment]
OK, I'm actually pretty happy with the results, although they are not ideal yet.

I made two changes. Now when you press Ctrl-Break, only the currently active Factor instance will be interrupted. No more of that global stuff. Plus, I've disabled the window ghosting, so no more of that switching to other apps with similar privileges and Ctrl-Breaking from there. On the down side, we don't get that "Not Responding" title and the ability to drag, minimize or close the hung app with the X button.

The second change is the fep queueing. I use that now, so the crash location is not as random as it used to be. But in my implementation we skip the console debugger completely. That was the whole point of the exercise - to interrupt things without having to put up with the console window lurking in the background. Instead, I do the fep queue, and when it arrives, I throw the same thing the console "t" command throws. So essentially pressing Ctrl-Break in the GUI window is now exactly the same as pressing Ctrl-C in console, followed immediately by the "t" command.

And this brings me to the less than ideal part of the matter. Throwing "t" is unsafe. Chances are you'll crash Factor, it doesn't matter that you've queued the fep, it can still crash on the "t". The good news is that when a tight loop is running, you will (almost?) always interrupt the loop and return to the Listener. Maybe that's true of any code (not just tight loops) you run from the Listener - if it's taking too long, you can interrupt it - because of the exception handler that the Listener wraps around user commands. The bad news is that if the Listener is idle and not having any problems, and you press Ctrl-Break, you will most certainly crash it. I'm not sure if that's the Listener's fault for wrapping nothing but entered quotation with exception catchers, or what is the deal there, but here we are. Don't press Ctrl-Break, unless you are sure the thing ain't coming back to life on its own: you'll kill it for sure.

If you have any ideas on how to throw the "t" more safely, please share! Maybe we could come up with a way to ignore the Ctrl-Break at all times except when the Listener runs a user quotation? That would be ideal, actually. If you have any pointers on how to do that, please tell me here or publish a patch I could try out.

Thank you, and have a good night.

**Comment by AlexIljin** [comment]
I DID IT!

I have reached the ideal state. The Listener now only enables the Ctrl-Break handler when it parses and runs user code. So now you can't terminate the Listener itself by accident.

Loading a vocab with an infinite loop in it - can be interrupted. Starting an infinite loop in the Listener - can be interrupted. Only works for the active/focused Factor instance/window. Theoretically you can still hit the GC or some other sensitive spot, but chances are pretty low, I think. Maybe some real-life testing will show that more safeguards are needed in choosing the moment of the "fep queuing". But that's fine-tuning that can be done by the VM gurus. My mission was to provide the mechanism, and it works.

I'm not sure about my placement of things in the sources, please feel free to relocate/rearrange them as you see fit. For instance, I've added two new primitives: enable-ctrl-break and disable-ctrl-break (similar to enable-gc-events and disable-gc-events). I put those new primitives in the "system" category, which I felt was rather ambitious, but I didn't find a better place, and did not have the courage to add a new category. Plus, I wanted to write some help about them, but did not find where it should go. A hint, please?

I have updated the pull request. You'll have to make new boot.*.images with the new primitives. I did not check if it all even compiles on other platforms, only tested on Windows. Probably some stubs are needed.

I knew I could do it, and I did it! Feels good.

**Comment by AlexIljin** [comment]
Forgot to mention that the Ctrl-Break mechanism is different from having the console open and doing the Ctrl-C + "t" in it. The difference is that the Listener is aware of it. When you do "t", you may crash anything in the VM. When you Ctrl-Break, you return to a safe point in the Listener.

**Comment by bjourne** [comment]
Good work! I can review the code you have written in a little while.

**Comment by mrjbq7** [comment]
Sweet! How does it know to only break Listener code and not background threads? Or does it always just interrupt into an open listener? Have you tested with multiple listener windows open to see it interrupts the correct one?

**Comment by AlexIljin** [comment]
@mrjbq7: I'm not sure what you mean by background threads. There are two threads in the Factor process: the VM thread and the keyboard watcher thread. When the watcher thread senses Ctrl-Break, it calls GetForegroundWindow to see which window in the system has focus. Then it finds out if that window belongs to the VM thread or not.

I didn't try starting multiple Listeners in a single Factor instance, it didn't occur to me (is that even possible?). In any case, only the one currently running user code will be interrupted, because it's the one that has the "recover" word with the exception handler on the stack of the VM thread.

You you were asking about background Factor processes, I answered that above. If you were asking about the cooperative threads, they can't be interrupted at the moment. If they have a "recover" point, we could trap them to return to that point. Could you create an example of an infinitely looping cooperative thread for me? What typically happens if such a thread throws an exception?

**Comment by mrjbq7** [comment]
I'm talking about Factor threads of which there are several.  Try `threads.` in the listener. They likely all run inside the single VM thread at the moment.

**Comment by mrjbq7** [comment]
Also, yes you can open several listener Windows, I believe using `open-listener-window`, but I'm not in front of my computer to test.

**Comment by mrjbq7** [comment]
I mean `listener-window`.

**Comment by bjourne** [comment]
Maybe the pr isn't as bad for performance as I thought. If you have a multicore system, perhaps the ctrl+break thread is ran on the other core so that the main factor thread isn't affected? Have you tried benchmarking with something slow like `[ 15 nsieve-main ] time`?

Polling the key state feels very wrong to me. It would be better if there was a way to have the thread sleep continually until waken up by a key press event. But maybe that's not possible to implement and if it doesn't affect performance it is no big deal.

What happens if the key is pressed only during the Sleep() call? Is it then lost? If not, we could increase the Sleep duration to 500 ms or so to reduce the performance impact.

Can the thread you have created be used to trap other keypresses like ctrl+c and can it do it if factor is running in a console window? If so, great! Then we can unify your code with the old ctrl+c-handling. That's much better than having two separate code paths for different keypresses.

I'm thinking that we could overload ctrl+c so that in the console it works as it does today. But with in a listener window, it works like your pr. Or the primitives you have added could be used to toggle the behaviour, independently of if you are running code in the console or not. Eg.

    run something slow <ctrl+c/break>
    ... low level debugger
    call `enable-ctrl-break`
    run something slow <ctrl+c/break>
    ... interrupt exception thrown

Because the feature you have implemented would of course be nice to have on Linux and Mac too.

As for safepoint enqueuing, it should be safe. As long as you have a `recover` block around the code that throws the interrupt exception, factor should have no trouble continuing running. I guess somewhere in `listener-step` in the `listener` vocab you need extra interrupted exception handling.

Pr looks really good though.

**Comment by jonenst** [comment]
What about non busy loops ?
That is, can you test and report the behavior of interrupting the following snippets in the listener ?
```
[ 1 seconds sleep t ] loop ! this one is yielding and resuming
```
and
```
10000000 seconds sleep ! this one is yielding and never resumes
```

**Comment by AlexIljin** [comment]
@jonenst: both of your examples leave the Listener responsive, but pressing Ctrl-Break terminates the application. I suppose that's because the Listener's exception handler is no longer on the call stack.

What is the expected behavior of a cooperative thread throwing an exception? Is there someone watching the cooperative threads and terminates them on error? If not, that's the problem. In a robust system I would expect cooperative threads throwing stuff to not bring down the rest of the system.

If we want to be able to terminate such threads and make them return control to the Listener, we need some support from the `threads` vocab.

**Comment by mrjbq7** [comment]
Like a global error handler?  Then the listener could display errors from background threads which would be a nice feature.

I had thought we log thread errors to the console if they occurred rather than crashing Factor.

**Comment by AlexIljin** [comment]
@bjourne: when the main thread is in a busy loop, the ctrl-break thread definitely runs on another core/hyperthread. I don't have a single-core PC around to run a benchmark that would give meaningful results. In any case, whenever one thread is busy, all the other threads are scheduled to other cores by the OS. One can't possibly interfere with another.

Again, we are talking about one API function call every 10-15 msec. That's negligible. OTOH, we could raise that to 500 msec and expect user to hold the Ctrl-Break combo for half a second. Or compromise on 100 msec, which is plenty from the usability perspective. The 10 msec would be necessary if we weren't sure if we can interrupt the main thread on the first attempt. That way we would keep trying every 10 msec until it's out of GC or whatever. But now that we have the fep queue thing going, which guarantees interruption, I would have no objection to up the delay to 100 msec.

[GetAsyncKeyState](https://msdn.microsoft.com/ru-ru/library/windows/desktop/ms646293%28v=vs.85%29.aspx) polls the keyboard state the moment it is called. If Ctrl-Break is pressed and released entirely within the Sleep() call, it will be lost.

I'll have to do some more research on the Ctrl-C suggestion. But one thing I can say right now: having Ctrl-C do anything but "copy selection to clipboard" on Windows is a bad idea. Not sure about Linuxes or Macs.

Do Macs have the Break button on their keyboards?

And yes, I did add the `cleanup` call to the Listener step for exactly that reason - to provide a known recovery point for the interrupt exception.

**Comment by mrjbq7** [comment]
One problem with frequent polling is increased battery usage and lack of deep sleep on laptop CPUs.

**Comment by AlexIljin** [comment]
@mrjbq7: this test
`[ 5 seconds sleep 5 throw t ] loop`
shows that exceptions in the threads don's bring the system down.

If I press Ctrl-Break while that thread is within its first 5 minutes, the ctrl-break handler is activated by the Listener, so I know that `listener-step` is not over. However, when my simulated "t" hits the VM, who is running while everyone sleeps?

What is running in the background when all threads are `sleep`ing or waiting for each other?

**Comment by AlexIljin** [comment]
@mrjbq7: we can pause the thread when it's not needed, and resume it in factor_vm::primitive_enable_ctrl_break.

**Comment by AlexIljin** [comment]
@bjourne: a bit of research into the Ctrl-Break handling shows that it can't be unified with Ctrl-C. Not on Windows, anyway. We have the ctrl-c handling provided by the OS, but on Windows you have two options to interrupt a console process: ctrl-c and ctrl-break. You can handle the first one by providing a callback (this is what VM does to drop into the debugger), but you can't handle the second.

So, if Factor is running with a console, Ctrl-Breaking in the Listener is handled by my thread, but Ctrl-Breaking in the console window is handled by the cmd.exe - it kills the process. The ctrl-c handler is never called if you ctrl-break the console.

Regarding the request to sleep more and wake up on keyboard events. There is this [RegisterHotKey](https://msdn.microsoft.com/ru-ru/library/windows/desktop/ms646309%28v=vs.85%29.aspx) function, but I've never used it. From the documentation it seems that registering the Ctrl-Break hotkey system-wide would limit us to a single Factor instance being interruptible by that. So if you hang the wrong one, you're out of luck. I'll have to experiment a little and report back.

**Comment by AlexIljin** [comment]
RegisterHotKey is no good: it is global, so only one process will be notified, and if it's already taken by someone, RegisterHotKey will fail.

**Comment by bjourne** [comment]
@jonenst That is because when you call `primitive_enable_ctrl_break` a different thread is active than when the safepoint that throws the interrupted exception occurs. It is the `factor_vm.ctx` variable that points to the current execution thread. So I think the answer is "don't do that" because the vm has no facility to interrupt a specific thread, only the currently executing one.

**Comment by bjourne** [comment]
Pr runs kind of fine on my Windows machine. There is some weird interaction between the ctrl+c handler and the added ctrl+break handler, but I think that can be fixed. What I think is left is:

1. Deleting the thread when `disable-ctrl-break` is called. That matches how the sampling profiler system works which creates/deletes threads in `start/end_sampling_profiler_timer`.

2. I'd like it to work so that when `enable-ctrl-break` is active, both pressing ctrl+break and ctrl+c in the console skips the low-level debugger. That makes the logic simpler and I think will make the ctrl+break feature easier to implement on Linux and Mac.


3. Benchmark. I'm gonna run that myself in a little while. If the ctrl+break does not increase the runtime of say `[ 15 nsieve-main ] time` by more than 20% on a multicore cpu then that is completely fine imho.

**Comment by mrjbq7** [comment]
Do we remove these handlers in deployed binaries? 

Is the Ctrl-Break functionality enabled always or just when the ui.tools is run?

**Comment by AlexIljin** [comment]
Only the Listener enables it for the duration of user code. Otherwise it's off.

**Comment by jonenst** [comment]
> Benchmark. I'm gonna run that myself in a little while. If the ctrl+break does not increase the runtime of say [ 15 nsieve-main ] time by more than 20% on a multicore cpu then that is completely fine imho.

I think that 20% is a high price to pay for something that you can easily achieve with SIGINT...

**Comment by AlexIljin** [comment]
@bjourne: number 2 is not possible on Windows AFAIK. For the reasons I explained earlier.
See also here: [CTRL+BREAK is always treated as a signal, but an application can change the default CTRL+C behavior](https://msdn.microsoft.com/ru-ru/library/windows/desktop/ms682541%28v=vs.85%29.aspx).

**Comment by AlexIljin** [comment]
I'm not sure how to correctly delete the thread when disable-ctrl-break is called. What if disable-ctrl-break is "immediately" followed by the enable-ctrl-break (where "immediately" is within the Sleep of the thread). It will non see that stop_on_ctrl_break was briefly set to false and then back to true again, so it will continue to run. But the primitive_disable_ctrl_break will have closed the thread handle, followed by the primitive_enable_ctrl_break creating a new one. So we end up with two threads treading on each other's toes.

OTOH if we wait for the previous thread to terminate (as is done with the sampler_thread), then the disable-ctrl-break call will on average take 50 msec to complete (assuming the Sleep(100) in the thread loop), which is quite a bit. IMO it's better to leave the thread around, let it check the one bool flag every 100 msec and go back to sleep, than to waste time on every disable-ctrl-break call. Mind you, there are two such calls per user code invocation (i.e. it runs twice for every time you hit Return in the Listener). For profiling that's fine, we are in the debugging mode anyway, but here - it's just sluggishness of the regular UI interactions.

Hey, I know! I'll make a synchronization object. The ctrl_break will wait on that object with its current Sleep time set as the timeout. If the object is signaled, the thread shall terminate. Otherwise it'll poll the keyboard after the timeout and go again.

**Comment by mrjbq7** [comment]
We are starting and stopping the thread each listener input return?

**Comment by bjourne** [comment]
Can the termination be implemented in the same way as for the sampling profiler? See the `factor_vm::end_sampling_profiler_timer` method. 

The problem with starting and stopping the thread could maybe be solved by wrapping the call to `(listener)` in a `with-ctrl-break` combinator. Then it would be running all the time while the listener is running and it wouldn't matter if `disable-ctrl-break` was a little slow.

**Comment by mrjbq7** [comment]
You can have multiple listeners per Factor process, so having it per-listener but also per-VM seems a little strange.

**Comment by AlexIljin** [comment]
@mrjbq7: "We are starting and stopping the thread each listener input return?"
- Could you rephrase that? I did not understand the question.

**Comment by mrjbq7** [comment]
I mean that in ``parse-lines-interactive`` you enable and disable the ctrl-break.  That gets called every time you hit enter to provide listener inputs... why not have it running for the duration of the program?

**Comment by AlexIljin** [comment]
@mrjbq7: Ah, I understand now. There are four reasons for not having the Listener wrapped the whole time:
a. we don't want to break out of the Listener itself, we only want to break out of the user code if it is running, and
II) for the same reason breaking out of `10000000 seconds sleep` crashes the system: we don't want to interrupt a random background thread while the Listener awaits another keypress, and
3 - end of the list of reasons.

This was also the answer to @bjourne's suggestion to wrap `(listener)` in a `with-ctrl-break` combinator. If we had a per-Factor-thread interruption support, then by all means we could do that. As it is now, we'll just crash the application if there is no tight loop to keep the CPU busy in one place.

**Comment by mrjbq7** [comment]
But background threads can still run during user input, so you just have a less likely chance of crashing?

Does the ctrl break thread require restarting after it throws? If not having the listener just catch that error and recover should be fine. No?

**Comment by AlexIljin** [comment]
@mrjbq7: with current implementation (enabling and disabling ctrl-break as needed) there is **no chance** of a random crash. If you extend the scope of the `with-ctrl-break` wrap and include a `sleep` or another form of `yield` in it, there **will be** a chance for some background thread to be active when user presses Ctrl-Break, which will result in a crash.

The ctrl-break thread **does not require** to be restarted after it throws. In fact, that's my preferred implementation: have it running all the time, and only handle Ctrl-Break when needed, ignore otherwise (a single boolean variable test). @bjourne asked me to change that to the current implementation, where the thread is created and destroyed every time it's needed, i.e. on every Enter key press in the Listener.

I am currently investigating the possibility to employ (Raw Input)[https://msdn.microsoft.com/en-us/library/windows/desktop/ms645536%28v=vs.85%29.aspx] so that the ctrl-break thread would only be activated when a WM_INPUT is posted to its message queue, which would obsolete the need for polling. I hope that would also obsolete the objection to keeping it around, because I dislike the amount of added code for creating/destroying the thread every time it is used.

**Comment by mrjbq7** [comment]
There is certainly a chance of crashing because "user code" could call a yield and in many cases does (socket networking code for example).

**Comment by AlexIljin** [comment]
@mrjbq7: ah, well, that's the caveat. It only works for tight loops, as was established earlier, until there is some kind of support from the `threads` vocab. I keep answering this question of yours: "why not have it running for the duration of the program?" And the answer is - it only works for tight loops.

**Comment by AlexIljin** [comment]
@mrjbq7: this is confusing : ))
If you were asking "why not have the ctrl-break thread running all the time, and have it monitor the keyboard and throw interrupts when needed, sleep otherwise", then I completely agree. That's the way was implemented in the PR until @bjourne asked to terminate the thread the way it's done with the sampler_thread. See his post above (in shows "3 days ago" for me now) with three items, item number 1.

**Comment by jonenst** [comment]
With a thread polling every few hundred milliseconds on keyboard events, it
would be disastrous to have it run any longer than necessary... so
creating/destroying everytime is a good thing.
Le 4 juin 2016 4:58 PM, "Alexander Iljin" <notifications@github.com> a
écrit :

**Comment by mrjbq7** [comment]
I agree with that, I guess I'm late to the party and thought there should be a way to be an OS way to be notified for Ctrl-Break instead of having to poll so frequently.

**Comment by bjourne** [comment]
I claim that it is *fine* for ctrl+break to crash the listener if no infinite loop is running. We say that the ctrl+break feature is for adults and say "don't do that". There is no way to ensure that ctrl+break *never* crashes Factor by interrupting an unlucky background thread that happens to be running unless the enqueue fep system is changed to becoming thread aware. 

When you call `enable-ctrl-break` and `disable-ctrl-break` is not so important to me. But it is important that the sequence `enable-ctrl-break disable-ctrl-break` leaves the vm "as it was before" and doesn't leave a running thread behind.

**Comment by AlexIljin** [comment]
@mrjbq7: I'm working on a no-polling solution.

**Comment by AlexIljin** [comment]
The no-polling solution is ready for testing, published in my repo. It works perfectly, but the code is way more involved.

One of the problems with Raw Input is that it's application-level. As soon as you call enable-ctrl-break, **all** keyboard input goes to the ctrl-break thread until you call disable-ctrl-break. I mean, you can still work with _other_ applications, but not with Listener or any of the main thread windows. This puts a cross on the idea of having the watcher thread permanently running in the background.

**Comment by AlexIljin** [comment]
@bjourne: did you benchmarks on the polling solution produce any results?

**Comment by bjourne** [comment]
Yes, and sorry for not reporting those. The polling had no discernible impact at all on performance. It didn't matter if I set the sleep to 500 ms, 100 ms, 30 ms or 1 ms. It's an Intel 5Y10 with four cores so the result is consistent with the theory that the polling thread runs on one core and doesn't interrupt what the main Factor thread is doing.

**Comment by AlexIljin** [comment]
@bjourne: thanks for the confirmation!

**Comment by AlexIljin** [comment]
@jonenst asked about the non-busy loops like `10000000 seconds sleep`, in which the Listener can get stuck. Back when I first implemented the c-break watcher such code would make the Listener fragile: pressing c-break would crash the VM. For some reason it is not the case now: the watcher thread is not active when such sleeping code is being executed by the Listener. Why is that so? I don't understand what happened.

I wanted to fix that remaining special case, but can no longer reproduce it, although I remember it being a real issue.

**Comment by bjourne** [comment]
Intuitively it should still crash because nothing major has changed. If no other threads wants to run, the io thread waits in a call to `GetQueuedCompletionStatus` for a number of milliseconds. When that is interrupted it should bring the thread and the Factor process down because there is nothing catching the exception. I can see if I can still reproduce it on Windows.

**Comment by bjourne** [comment]
If you type `exit` with ctrl+break on in the listener, Factor doesn't exit properly. I think it is because Factor tries to terminate before the `disable-ctrl-break` primitive is called. So the ctrl+break thread is keeping the Factor process alive.

**Comment by AlexIljin** [comment]
Both 32- and 64-bit nightly builds from "2018-01-16 17-13" exit properly and cleanly if you enter `exit` in the listener with ctrl-break handling enabled (`USING: namespaces listener ; t handle-ctrl-break set-global`).
